Add first questions and answers (#1)
CTSRD-CHERI/cheri-faq: src/QUESTIONS.md

## General

### What CHERI is?

See
[An Introduction to CHERI](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-941.pdf).

### What CHERI-extended hardware is available to use?

[Arm Morello](https://www.arm.com/architecture/cpu/morello)
is currently the only System-on-Chip with a CHERI-extended CPU.

There are several
[CHERI-RISC-V prototypes on FPGA](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-risc-v.html),

> including a superscalar, out-of-order Toooba-based core.

Reviewer: It might make sense to mention RiscyOO here as well since more people might be aware of it and https://github.com/bluespec/Toooba says it's a lightly modified version there of

Reviewer 2: Agreed, there is also a paper on RiscyOO(https://people.csail.mit.edu/szzhang/paper/micro2018.pdf), but not documentation on Toooba other than a few READMEs. If we want to be precise here, then we should say that Toooba is based RiscyOO and then we took it and extended it for CHERI resulting in CHERI-Toooba. So they are three different things.

Additionally, CHERI-RISC-V and Morello can be emulated using QEMU-CHERI or
Morello FVP.

### How can I get an Arm Morello board?

Go to the
[Digital Security by Design](https://www.dsbd.tech/get-involved/morello-board-request/)
website and follow the instructions there.

### How can I emulate CHERI?

Go to the
[cheribuild](https://github.com/CTSRD-CHERI/cheribuild)

> repository and follow the instructions there.

Reviewer: The instructions are rather long, I wonder if we should provide some succinct guidance here.

  for
  [Arm Morello](https://www.arm.com/architecture/cpu/morello)

* [CHERIoT RTOS](https://github.com/microsoft/cheriot-rtos) for the
  [CHERIoT platform](https://aka.ms/cheriot-tech-report)

* seL4

  There are several institutions interested in developing seL4 for CHERI.
  You can talk to them in the `#sel4` channel at the
  [CHERI-CPU](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-slack.html)
  Slack.

### Which operating system should I use to experiment with CHERI?

Check the
[CHERI OS-feature matrix](https://www.morello-project.org/cheri-feature-matrix/)
to compare CheriBSD and Morello Linux.

If you want to work on a project that includes multiple third-party software
dependencies, you probably want to use
[CheriBSD](https://cheribsd.org/)
as it provides pre-compiled software that can be installed with package managers.

### How do I start with CheriBSD/Morello?

Follow the
[Getting Started with CheriBSD guide](https://ctsrd-cheri.github.io/cheribsd-getting-started/)
and do not skip any section.

The guide has been structured to make you aware of actions you should take
before trying to use CheriBSD.

> For example, you should upgrade your Morello board firmware before installing

Reviewer: ```suggestion
Of critical importance, you should upgrade your Morello board firmware before installing
```

specification for more information on the ABIs.

### Should I compile my code for the pure-capability or hybrid ABI?

It might seem that compiling code for the hybrid ABI is a better approach to
port software to CHERI than compiling it for the pure-capability ABI because you
can incrementally introduce CHERI capabilities in your code, and not worry about
all incompatibilities with CHERI at once.

However, when using the hybrid ABI, you have to annotate all pointers as CHERI
capabilities, and still you will not have any guarantees that there are no
legacy pointers left.
Additionally, all third-party libraries that your code depends on will not use
CHERI capabilities, unless you change them as well.

There are still some situations when choosing the hybrid ABI over the
pure-capability ABI has benefits.
In case of extremely large code bases (e.g., the Chromium browser and its
dependencies), you might want to compartmentalize parts of your code base now
(e.g., the V8 JavaScript engine in Chromium), parts that are known for security
vulnerabilities with significant impacts, rather than spend a lot of time
trying to port the whole project for the pure-capability ABI and then

> compartmentalize it.

Reviewer: Do we believe this is a good example? I'm skeptical on the grounds that we've been fighting with V8 for many months.

Author: AFAIK @gcjenkinson has been working on CheriABI V8 so far, and didn't get to the stage to try to link it somehow with Chromium. Although, I'm not sure if that's the plan for a hybrid ABI Chromium with memory-safe V8.

I'd be fine with mentioning another example if we can think of one. I'd be tempted to mention compartmentalizing binaries with closed source code but that raises other issues.

In most cases, porting software to the pure-capability ABI is straightforward
and does not heavily disrupt code.
For example, the Capabilities Limited company in their report
[Assessing the Viability of an Open-Source CHERI Desktop Software Ecosystem](https://www.capabilitieslimited.co.uk/_files/ugd/f4d681_e0f23245dace466297f20a0dbd22d371.pdf)
shows that they needed to modify 0.026% lines of code in around 6 million lines
of code from KDE, Qt, X11 for CheriABI to run KDE Plasma on top of Xvnc on

> CheriBSD.

Reviewer: I think I'd put this paragraph first.

KDE Plasma 5, Git, Rsync, sudo, Bash, tmux.
Besides that, there are ~24,000 hybrid ABI packages, including LLVM-CHERI,
GDB-CHERI, Chromium, Firefox, Vim, Python, Meson, Ninja.

You can list all available software on CheriBSD with `pkg64c rquery %n` and
`pkg64 rquery %n`.
Alternatively, you can search for a specific package using
`pkg64c search <pattern>` and `pkg64 search <pattern>`.

The packages are built using
[CheriBSD ports](https://github.com/CTSRD-CHERI/cheribsd-ports),
a fork of
[FreeBSD ports](https://github.com/freebsd/freebsd-ports)
that includes a collection of ~32,000 third-party software ports for FreeBSD.

See the
[Getting Started with CheriBSD](https://ctsrd-cheri.github.io/cheribsd-getting-started/packages/)
guide for more information on third-party software packages,
and the
[CheriBSD Ports and Packages](https://freebsdfoundation.org/wp-content/uploads/2023/05/CheriBSD_ports.pdf)
article to read more on the package building process.

### How can I print a CHERI capability in CHERI C/C++?

See the
[Displaying Capabilities](https://github.com/CTSRD-CHERI/cheri-c-programming/wiki/Displaying-Capabilities)

> wiki page for printf(3) string formats.

Reviewer: Maybe also a reference to https://man.cheribsd.org/cgi-bin/man.cgi/dev/strfcap?

  There are plans to make this model available for Morello as well.

## CheriBSD

### How can I upgrade CheriBSD?

Currently, there are no binary upgrades for CheriBSD.

The only way to upgrade a CheriBSD host is to build and install CheriBSD from
source code.
See the
[wiki page](https://github.com/CTSRD-CHERI/cheripedia/wiki/HOWTO:-Build-CheriBSD-natively-on-Morello)
for instructions how to do this.
Note that we cannot guarantee stability when upgrading CheriBSD and it is
important to make sure you can recover your data if an upgrade fails.

### How can I build CheriBSD from source code?

You have two choices:
* Build and install CheriBSD natively on Arm Morello using
[these instructions](https://github.com/CTSRD-CHERI/cheripedia/wiki/HOWTO:-Build-CheriBSD-natively-on-Morello)

* Cross-compile (on FreeBSD, macOS or Linux) CheriBSD for CHERI-RISC-V or Arm Morello using
[cheribuild](https://github.com/CTSRD-CHERI/cheribuild)

### Does CheriBSD implement spatial safety (e.g., to prevent out-of-bounds bugs)?

The official CheriBSD release runs basic programs and libraries compiled
for the pure-capability ABI and a hybrid kernel.

It also includes a pure-capability kernel in

> `/boot/kernel.GENERIC-MORELLO-PURECAP`.

Reviewer: I wonder if this wants to reference the how to select a kernel question?

## General

> ### What CHERI is?

Reviewer 3: What is CHERI?

1. Build a software stack and run CheriBSD in a QEMU-based VM.

   * For Arm Morello: `cheribuild.py run-morello-purecap -d`

   * For CHERI-RISC-V: `cheribuild.py run-riscv64-purecap -d`

## Operating Systems

### What operating systems are being developed for CHERI?

* [CheriBSD](https://cheribsd.org/) for
  [Arm Morello](https://www.arm.com/architecture/cpu/morello) and
  [CHERI-RISC-V](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-risc-v.html)

* [Linux](https://linux.morello-project.org/docs/) for
  [Arm Morello](https://www.arm.com/architecture/cpu/morello)

* [Android](https://git.morello-project.org/morello/docs/-/blob/morello/mainline/android/readme.rst)
  for
  [Arm Morello](https://www.arm.com/architecture/cpu/morello)

* [CHERIoT RTOS](https://github.com/microsoft/cheriot-rtos) for the
  [CHERIoT platform](https://aka.ms/cheriot-tech-report)

* seL4

  There are several institutions interested in developing seL4 for CHERI.
  You can talk to them in the `#sel4` channel at the
  [CHERI-CPU](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-slack.html)
  Slack.

### Which operating system should I use to experiment with CHERI?

> 

Reviewer 3: At the moment CheriBSD is the most fully featured operating system running on CPUs with CHERI extensions.  If you intend to work with code that runs on UNIX-like operating systems we recommend you start with CheriBSD.

Author: I've added it with minor differences.

Check the
[CHERI OS-feature matrix](https://www.morello-project.org/cheri-feature-matrix/)
to compare CheriBSD and Morello Linux.

If you want to work on a project that includes multiple third-party software
dependencies, you probably want to use
[CheriBSD](https://cheribsd.org/)
as it provides pre-compiled software that can be installed with package managers.

> ### How do I start with CheriBSD/Morello?

Reviewer 3: How do I get started with CheriBSD on Morello?

Follow the
[Getting Started with CheriBSD guide](https://ctsrd-cheri.github.io/cheribsd-getting-started/)
and do not skip any section.

The guide has been structured to make you aware of actions you should take
before trying to use CheriBSD.
Of critical importance, you should upgrade your Morello board firmware before
installing CheriBSD.

## Software porting

### What programming languages can I use on CHERI?

* CHERI C/C++

>   Currently, the only mature programming languages for CHERI are CHERI C and

Reviewer 3: the most mature programming...

  CHERI C++.

  See the
  [CHERI C/C++ Programming guide](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-947.pdf)
  for more information.

* Rust

  There is an
  [ongoing project](https://soft-dev.org/events/cheritech22/slides/Cooksey.pdf)
  at the University of Kent to add support for CHERI in Rust.

* Python/Ruby/Lua/JavaScript/...

  There are no interpreters that support CHERI for these programming languages.

### Can I use Python with CHERI?

There is no Python interpreter that would run your Python code using
capability-aware instructions.

You can use a Python interpreter compiled for a baseline architecture but it

> does not give you any security benefits.

Reviewer 3: does not give you any of the security benefits of a CHERI enabled CPU.
